lib/testutils/src/lib.rs

```diff
@@ -340,6 +339,88 @@ impl TestWorkspace {
     }
 }
 
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub enum DirEntry {
+    File { contents: Vec<u8>, executable: bool },
+    Symlink { target: String },
+    GitSubmodule { commit: CommitId },
+}
+
+impl DirEntry {
+    pub fn empty_file() -> Self {
+        DirEntry::File {
+            contents: vec![],
+            executable: false,
+        }
+    }
+
+    pub fn text_file(contents: &str) -> Self {
+        DirEntry::File {
+            contents: contents.as_bytes().to_vec(),
+            executable: false,
+        }
+    }
+
+    pub fn binary_file(contents: Vec<u8>) -> Self {
+        DirEntry::File {
+            contents,
+            executable: false,
+        }
+    }
+
+    pub fn symlink(target: &str) -> Self {
+        DirEntry::Symlink {
+            target: target.to_owned(),
+        }
+    }
+
+    pub fn submodule(commit: CommitId) -> Self {
+        DirEntry::GitSubmodule { commit }
+    }
+
+    pub fn executable(self, executable: bool) -> Self {
+        match self {
+            Self::File { contents, .. } => Self::File {
+                contents,
+                executable,
+            },
+            DirEntry::Symlink { .. } => panic!("symlinks cannot be executable"),
+            DirEntry::GitSubmodule { .. } => panic!("submodules cannot be executable"),
+        }
+    }
```

Comment:
To avoid these panics, and to simplify the API overall, I would make this a "static" function taking the content as input. We can rename `text_file()` to `normal_file()` then. Hopefully we only need one `executable(contents: &str) -> Self`, but adding a `binary_executable(contents: Vec<u8>)` if necessary also seems fine.

Reply:
The new builder API I've added works with a dedicated type for file entries, so it's not possible to call `executable()` on symlink or submodule entries. This serves to avoid these panics while still keeping file contents and executable bit separate.